pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -624,6 +646,89 @@ impl RemoteTimelineClient {
         Ok(())
     }
 
+    /// Set the deleted_at field in the remote index file.
+    ///
+    /// This fails if the upload queue has not been `stop()`ed.
+    ///
+    /// The caller is responsible for calling `stop()` AND for waiting
+    /// for any ongoing upload tasks to finish after `stop()` has succeeded.
+    /// Check method [`RemoteTimelineClient::stop`] for details.
+    pub(crate) async fn persist_index_part_with_deleted_flag(
+        self: &Arc<Self>,
+    ) -> anyhow::Result<()> {
+        let index_part_with_deleted_at = {
+            let mut locked = self.upload_queue.lock().unwrap();
+
+            // We must be in stopped state because otherwise
+            // we can have inprogress index part upload that can overwrite the file
+            // with missing is_deleted flag that we going to set below
+            let stopped = match &mut *locked {
+                UploadQueue::Uninitialized => anyhow::bail!("is not Stopped but Uninitialized"),
+                UploadQueue::Initialized(_) => anyhow::bail!("is not Stopped but Initialized"),
+                UploadQueue::Stopped(stopped) => stopped,
+            };
+
+            if let Some(deleted_at) = stopped.deleted_at.as_ref() {
+                anyhow::bail!("timeline is deleting, deleted_at: {:?}", deleted_at);
+            }
```

Comment:
I have https://github.com/neondatabase/neon/pull/4147 with another test case, also against this PR . Don't understand why that one doesn't fail due to `delete_at = Some()`.

Reply:
I think the problem is that here we **must** synchronize on the upload.

If we error out here, we cannot make progress after one has been cancelled *after* this method but on the next await point (or failpoint).

If we don't error here, there's no guarantee that anyone will get to retry the upload before files and local inmemory state has been removed.

So I think we are back to the original problem, what would the request coalescing help here: it would allow synchronizing and not care about these two tasks working on the same lockless algorithm.

Reply:
Trying the one spawned task to delete a timeline horror:
- need to fuzz test that we cannot have the serialization fail
    - the indexpart update is really annoying
    - wonder why we didn't go with an s3 marker file
- handling the `anyhow::Result<()>` is equally frustrating
    - initially tried with more of cloneable result approach but this is just frustrating
    - perhaps DeleteTimelineError could be just really simple

Reply:
> Don't understand why that one doesn't fail due to delete_at = Some().

it’s because the test does not configure remote storage

Reply:
Fixed it in PR #4147 commit https://github.com/neondatabase/neon/pull/4147/commits/0ab36758a260049e9c7839722514f5a18f8e5406